server/matchmaker/algorithm/team_matchmaker.py

```diff
@@ -0,0 +1,262 @@
+    """
+
+    def find(self, searches: Iterable[Search]) -> List[Match]:
+        searches = list(searches)
+        searches.sort(key=lambda s: s.average_rating, reverse=True)
+        self._logger.debug("=== starting matching algorithm ===")
+        possible_games = set()
+        for index, search in enumerate(searches):
+            self._logger.debug("building game for %s", repr(search))
+            participants = self._pick_neighboring_players(searches, index)
+            try:
+                match = self.make_teams(list(participants))
+                game = self.calculate_game_quality(match)
+                possible_games.add(game)
+            except AssertionError:
+                self._logger.warning("failed to assign even teams. Skipping this game...")
+        self._logger.debug("got %i games", len(possible_games))
+        for game in possible_games:
+            self._logger.debug("game: %s vs %s rating disparity: %i quality: %f",
+                               repr(game.match[0]),
+                               repr(game.match[1]),
+                               game.match[0].cumulated_rating - game.match[1].cumulated_rating,
+                               game.quality)
+        return self._pick_best_noncolliding_games(list(possible_games))
+
+    def _pick_neighboring_players(self, searches: List[Search], index: int) -> List[Search]:
+        participants = []
+        i = 0
+        number_of_players = 0
+        out_of_bounds_counter = 0
+        while number_of_players < self.team_size * 2 and out_of_bounds_counter < 2:
+            try:
+                candidate = searches[index]
+                out_of_bounds_counter = 0
+                if number_of_players + len(candidate.players) <= self.team_size * 2:
+                    participants.append(candidate)
+                    number_of_players += len(candidate.players)
+            except IndexError:
+                out_of_bounds_counter += 1
+            i += 1
+            index += i * pow(-1, i)
+        return participants
+
+    def make_teams(self, searches: List[Search]) -> Tuple[Search, Search]:
+        avg = CombinedSearch(*searches).average_rating
+        team_target_strength = CombinedSearch(*searches).cumulated_rating / 2
+        searches_dict = self._searches_by_size(searches)
+        team_a = []
+        team_b = []
+
+        if searches_dict[self.team_size]:
+            search = searches_dict[self.team_size].pop()
+            team_a.append(search)
+            searches.remove(search)
+        elif searches_dict[self.team_size - 1]:
+            search = searches_dict[self.team_size - 1].pop()
+            filler = self._find_most_balanced_filler(avg, search, searches_dict)
+            team_a.append(search)
+            team_a.append(filler)
+            searches.remove(search)
+            searches.remove(filler)
+        else:
+            team_a, searches = self.run_karmarkar_karp_algorithm(searches)
+        team_b.extend(searches)
```

Comment:
If you do it like this, then you don't need to call `remove`.
```suggestion
        team_b.extend(search for search in searches if search not in team_a)
```